### PR TITLE
[ADD][16.0] product_print_category

### PR DESCRIPTION
**Pull request by legalsylvain**
### Description

This module is designed to extend product printing features. It allows user to create new print categories of products depending of the data that are on the labels of the products variants.

When a field has changed, the product will be set as 'To print' automatically.

This module requires to install custom modules to have pricetags templates, or alternatively to create pricetag reports templates by UI.

### Configure

* Add yourself to the 'Print Category Manager' group in the 'Pricetags' category.

* Go to 'Sales' > 'Products' > 'Print Categories'

* Find or create new print categories of products
    * Set the fields that will trigger to print again the product label (e.g : name, sale price etc.)
    * Set the qweb view that will be used for this category

product_print_category_form

Alternatively, you can create product print categories and qweb in a custom module.

**To affect products to the categories**

* Go to "Sales" > "Products" > "Product variants" (or "Products" if variants are not activated).
* Go to your product in General Information tab
* Set a print category

product_product_form

**Create Print Category Rules**

* Go to 'Sales' > 'Products' > 'Print Category Rules'

Create various ordered rules depending on product category and company.

This system allow to automatically set / remove print category to a product, when changing its company or its category.

product_print_category_rule_tree

### Usage

* Go to "Sales" > "Products" > "Print Categories"
* Choose between "Print Obsolete Products" or "Print All Products"

product_print_wizard_form

* Alternatively you can select product in product or variants list view and click on "Action" > "Print Products"

CC :  @remytms I tried to backport this patch https://github.com/grap/grap-odoo-incubator/pull/149/files in V16 but it doesn't seem to work. any idea ?

### CODE REVIEW

**Comment by remytms** [comment]
@legalsylvain I seams to work on my local computer with a fresh 16.0 installed with demo-data.

This is the patch I applied before testing:
```
diff --git a/product_print_category/report/report_pricetag.xml b/product_print_category/report/report_pricetag.xml
index 1be2a8a..37b01f8 100644
--- a/product_print_category/report/report_pricetag.xml
+++ b/product_print_category/report/report_pricetag.xml
@@ -11,7 +11,7 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
             <t t-foreach="categories_data" t-as="category_data">
                 <div class="page">
                     <t
-                        t-call="#{ category_data['print_category'].qweb_view_id.xml_id }"
+                        t-call="#{ category_data['print_category'].qweb_view_id.key }"
                     />
                 </div>
             </t>
```

This is the PDF report I get: 
[Pricetags.pdf](https://github.com/OCA/product-attribute/files/9870510/Pricetags.pdf)

**Comment by remytms** [comment]
@legalsylvain Tell me which error you get when using my patch, if you encounter some. Or a way to reproduce it. :)

**Comment by legalsylvain** [comment]
> @legalsylvain Tell me which error you get when using my patch, if you encounter some. Or a way to reproduce it. :)

That's weird ! I tested again, and all seems to work properly. I added the last commit to backport your improvment.

Thanks ! 

Note : could you review the rest of the PR ?

**Comment by remytms** [comment]
For sure I can do a review. :)

File: product_print_category/README.rst
Comment (remytms) [inline product_print_category/README.rst]: I think that theses links to https://github.com/grap/grap-odoo-incubator should be replaced by the OCA one, has this module will be in OCA now. :)

File: product_print_category/__manifest__.py
Comment (remytms) [inline product_print_category/__manifest__.py]: Can you change this to `Coop IT Easy SC` (juridic name reform has occurred in Belgium).

**Review by remytms — CHANGES_REQUESTED** [review]
Read all the code and tested it with demo data. Just some naming changes requested.

File: product_print_category/README.rst
Comment (legalsylvain) [reply product_print_category/README.rst]: No need. This text is autogenerated by bot when merging. Only fragment should be reviewed

(but done) ;-)

File: product_print_category/__manifest__.py
Comment (legalsylvain) [reply product_print_category/__manifest__.py]: done.

File: product_print_category/README.rst
Comment (remytms) [reply product_print_category/README.rst]: Thanks for the info. :)

**Review by remytms — APPROVED** [review]

**Comment by legalsylvain** [comment]
@PierrickBrun could you review this one ?

**Review by PierrickBrun — COMMENTED** [review]
The product.template view is broken

Capture d’écran du 2022-12-16 16-31-35

Apart from that everything looks functional

**Comment by legalsylvain** [comment]
@PierrickBrun : I fixed the view issue. could you take a look, and update your review ? 
thanks !

**Review by PierrickBrun — APPROVED** [review]

**Comment by legalsylvain** [comment]
Please avoid to merge, I have a final change to do before.

**Comment by rousseldenis** [comment]
@legalsylvain Still WIP ?

**Comment by legalsylvain** [comment]
Hi @rousseldenis That's OK for me !

**Comment by remytms** [comment]
@legalsylvain Ping me if you need another review since your last changes. And if I understand correctly, this is not WIP anymore, so you can remove this from the title of this PR. :)

**Comment by legalsylvain** [comment]
@rousseldenis could you merge it ?

**Comment by rousseldenis** [comment]
@legalsylvain Could you just remove the merge commit ? Thanks

**Comment by legalsylvain** [comment]
> @legalsylvain Could you just remove the merge commit ? Thanks

done.

**Comment by legalsylvain** [comment]
Hi maintainers : could you merge it ?

**Comment by carmenbianca** [comment]
Kind ping @ maintainers to have a look at merging this :)

**Comment by legalsylvain** [comment]
@OCA/product-maintainers : this one can be merged.

**Comment by carmenbianca** [comment]
@OCA/product-maintainers this is still ready to be merged :tada:

**Comment by legalsylvain** [comment]
It should be OK now. could you retry to merge ?

**Comment by dreispt** [comment]
Sure! Shouldn't you be partying?!

/ocabot merge nobump

**Pull request merged**
